### PR TITLE
[DataLoader] Move BufferedShuffle from Dataset to DataPipe

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52141 [DataLoader] Move BufferedShuffle from Dataset to DataPipe**

Remove BufferShuffleDataSet, as it's not being used anywhere within PyTorch (no usage on Github based on a search) and it's not included in the release of PyTorch 1.7.1.

Differential Revision: [D26710940](https://our.internmc.facebook.com/intern/diff/D26710940)

### CODE REVIEW

**Comment by glaringlee** [comment]
Two questions here:
1. should we add an optional arg or an api to pass a seed into shuffle datapipe?
2. should we add a default buffer size?

**Review by glaringlee — COMMENTED** [review]
Two questions here:

should we add an optional arg or an api to pass a seed into shuffle datapipe?
should we add a default buffer size?

**Comment by ejguan** [comment]
> should we add an optional arg or an api to pass a seed into shuffle datapipe?

Considering our new architecture is designed as a pipeline, random seed is needed but needs to figure out the way to set it. As we can have a sequence of DataPipe required random seed, it's debatable for us where to put random seed. I writing a doc for it, but I can talk about it briefly here.
1. For each DataPipe required random seed, we need preserve it's own random state. The sequence of DataPipe will use the last set random seed as the seed for the following random operations. So, we need to preserve the state:
```py
class RandomStateDataPipe:
    def __init__(self, datapipe, seed):
        self.datapipe = datapipe
        self.seed = seed
    
    def get_iter(self):
        raise NotImplementedError
    
    def __iter__(self):
        it = get_iter() # The original generator
        random.seed(self.seed)
        old_state = random.getstate()
        while True:
            random.setstate(old_state)
            v = next(it)  # Random operation
            old_state = random.getstate() # Preserve the state after one iteration
            yield v

class ShuffleDataPipe(RandomStateDataPipe):
    def get_iter(self):
        ...
```
In this case, we definitely needs to add seed in the arguments.

2. The whole pipeline owns the random state. So we can set random seed 
```py
dp1 = FirstDataPipe(...)
dp2 = SecondDataPipe(dp1, ...) # Has random here
...
dpn = NthDataPipe(...)

# In DataLoader
class DataLoader:
    def reset_random(self):
        rank = dist.get_rank()
        random.seed(seed+rank)
        torch.manual_seed(seed+rank)
        ...
    def __iter__(self):
        reset_random() # Each epoch has same randomness
        for d in dpn:
            yield d
```

> should we add a default buffer size?

I don't think we should add a default `buffer_size`, as users need to determine it. If we set the default as 1, the behavior that it won't shuffle is not intuitive for them. As the size of datasource is not always available, I am not sure if there is a reasonable value for all data sources.

Besides, I also want to discuss about if we should add another argument like `reshuffle_per_iter` to change the random seed over each epoch.

**Comment by ejguan** [comment]
Notes:
- The signature remains same considering randomness is controlled by DataLoader
- Example and notes of random seeding for multiprocessing and multithreading should be finished after DL is done.

**Review by VitalyFedyunin — APPROVED** [review]
